tiledb/sm/cpp_api/query.h

```diff
@@ -395,6 +394,120 @@ class Query {
     return elements;
   }
 
+  /**
+   * Adds a 1D range along a subarray dimension, in the form
+   * (start, end, stride). The datatype of the range
+   * must be the same as the subarray domain type.
+   *
+   * **Example:**
+   *
+   * @code{.cpp}
+   * // Set a 1D range on dimension 0, assuming the domain type is int64.
+   * int64_t start = 10;
+   * int64_t end = 20;
+   * // Stride is optional
+   * subarray.add_range(0, start, end);
+   * @endcode
+   *
+   * @param dim_idx The index of the dimension to add the range to.
+   * @param range The range to add.
+   * @return Reference to this Subarray
+   */
+  Query& add_range(
+      uint32_t dim_idx,
+      const void* start,
+      const void* end,
+      const void* stride = nullptr) {
+    auto& ctx = ctx_.get();
+    ctx.handle_error(tiledb_query_add_range(
+        ctx.ptr().get(), query_.get(), dim_idx, start, end, stride));
+    return *this;
+  }
+
+  /**
+   * Retrieves the number of ranges for a given dimension.
+   *
+   * **Example:**
+   *
+   * @code{.cpp}
+   * unsigned dim_idx = 0;
+   * uint64_t range_num = query.range_num(dim_idx);
+   * @endcode
+   *
+   * @param attr_name The attribute name.
+   * @return The estimated size in bytes.
+   */
+  uint64_t range_num(unsigned dim_idx) const {
+    auto& ctx = ctx_.get();
+    uint64_t range_num;
+    ctx.handle_error(tiledb_query_get_range_num(
+        ctx.ptr().get(), query_.get(), dim_idx, &range_num));
+    return range_num;
+  }
+
+  std::array<const void*, 3> range(unsigned dim_idx, uint64_t range_idx) {
```

Comment:
similarly here, although this can come in a follow on PR

Reply:
Will address.